File: intake-soa/web/runs.js
/* Run log. Read-only view over outputs/runs.db — a record of what happened, never a way to
   change it. The two parts of the assignment share one row shape (produced / expected / flagged),
   so one table renders both and the header just counts them separately. */

const $ = (id) => document.getElementById(id);
const el = (tag, cls, text) => {
  const n = document.createElement(tag);
  if (cls) n.className = cls;
  if (text !== undefined) n.textContent = text;
  return n;
};

let all = [];
let part = "";            // "" = every part

const PARTS = { soa: "SoA extraction", agent: "Study build" };
const label = (p) => PARTS[p] || p;
const when = (iso) => new Date(iso).toLocaleString();

async function load() {
  let data;
  try {
    const res = await fetch("/api/runs");
    if (!res.ok) throw new Error(`${res.status} ${res.statusText}`);
    data = await res.json();
  } catch (e) {
    $("panel").replaceChildren(el("div", "err", `Could not read the run log: ${e.message}`));
    return;
  }
  all = data.runs;
  renderTotals(data.totals);
  renderFilters();
  render();
}

function renderTotals(totals) {
  const box = $("totals");
  box.replaceChildren();
  if (!totals.length) { box.append(el("p", "rail__note", "No runs recorded yet.")); return; }
  for (const t of totals) {
    const d = el("div", "doc");
    d.dataset.status = t.not_done ? "failed" : "done";
    d.append(el("div", "doc__name", label(t.part)));
    const bits = [`${t.runs} run${t.runs === 1 ? "" : "s"}`];
    if (t.expected) bits.push(`${t.produced}/${t.expected} produced`);
    else if (t.produced != null) bits.push(`${t.produced} produced`);
    if (t.flagged != null) bits.push(`${t.flagged} flagged`);
    if (t.avg_seconds != null) bits.push(`${t.avg_seconds}s avg`);
    if (t.not_done) bits.push(`${t.not_done} not done`);
    d.append(el("div", "doc__meta", bits.join(" · ")));
    box.append(d);
  }
}

function renderFilters() {
  const box = $("filters");
  box.replaceChildren();
  const parts = ["", ...new Set(all.map((r) => r.part))];
  for (const p of parts) {
    const b = el("button", `tab${p === part ? " on" : ""}`, p ? label(p) : "Everything");
    b.setAttribute("aria-selected", String(p === part));
    b.addEventListener("click", () => { part = p; renderFilters(); render(); });
    box.append(b);
  }
}

/** How much of what was asked for came out. Blank when the run had no target to compare against. */
function recall(r) {
  if (!r.expected) return r.produced == null ? "—" : String(r.produced);
  return `${r.produced}/${r.expected} · ${Math.round((r.produced / r.expected) * 100)}%`;
}

function render() {
  const rows = part ? all.filter((r) => r.part === part) : all;
  $("head").textContent = rows.length
    ? `${rows.length} run${rows.length === 1 ? "" : "s"}${part ? ` · ${label(part)}` : ""}`
    : "No runs";

  const panel = $("panel");
  panel.replaceChildren();
  if (!rows.length) {
    const e = el("div", "empty");
    e.append(el("b", null, "Nothing recorded yet"),
             el("span", null, "Extract a protocol, or ingest an agent bench run, then reload."));
    panel.append(e);
    return;
  }
  for (const r of rows) panel.append(runBlock(r));
}

function runBlock(r) {
  const block = el("div", "block");

  const head = el("div", "block__head");
  const h = el("h2", null, r.subject);
  head.append(h);
  const meta = el("p", null,
    [label(r.part), r.input, when(r.started), r.seconds != null ? `${r.seconds}s` : null]
      .filter(Boolean).join(" · "));
  head.append(meta);
  head.append(el("span", `pill pill--${r.status === "done" ? (r.flagged ? "info" : "ok") : "flag"}`,
    r.status === "done" && r.flagged ? `${r.flagged} need a person` : r.status));
  block.append(head);

  const body = el("div", "block__body");
  const dl = el("dl", "rev");
  const add = (k, v, flag) => {
    dl.append(el("dt", null, k));
    const dd = el("dd", null, v);
    if (flag) dd.style.color = "var(--flag)";
    dl.append(dd);
  };

  add(r.part === "agent" ? "fields built" : "schedules found", recall(r),
      !!r.expected && r.produced < r.expected);
  add("needs a person", String(r.flagged ?? 0), (r.flagged ?? 0) > 0);

  const d = r.detail || {};
  if (r.part === "agent") {
    if (d.visits) add("visits", d.visits, d.visits.split("/")[0] !== d.visits.split("/")[1]);
    if (d.forms) add("forms", d.forms, d.forms.split("/")[0] !== d.forms.split("/")[1]);
    if (d.panel_status) add("agent reported", d.panel_status);
    if (d.ledger_states) add("ledger", Object.entries(d.ledger_states).map(([k, n]) => `${k} ${n}`).join(" · "));
  } else {
    const doc = d.document || {};
    if (doc.pages) add("document", `${doc.pages} pages · ${doc.tables} tables · ${doc.convert_seconds}s to convert`);
    if (d.model) add("model", d.model);
    for (const s of d.schedules || []) {
      add(`${s.soa_id}`, `pages ${(s.pages || []).join(", ")} · ${s.fragments} fragment(s) · ${s.footnotes} footnote(s)`);
      const un = s.review?.footnotes_never_used_in_table || [];
      const orph = s.review?.markers_in_table_without_definition || [];
      if (un.length) add(`${s.soa_id} — footnotes with no target`, un.join(", "), true);
      if (orph.length) add(`${s.soa_id} — markers with no definition`, orph.join(", "), true);
    }
    if (d.error) add("error", d.error, true);
  }
  body.append(dl);

  // The escalated path, spelled out: what the agent was blocked on, what it already tried, and
  // how many plan items each card is holding up. This is what a person acts on.
  for (const c of d.gate_cards || []) {
    const card = el("details", "raw");
    card.append(el("summary", null, `${c.signature} — ${c.question}`));
    const pre = el("pre", null, [c.blocks, c.tried].filter(Boolean).join("\n\n") || "—");
    card.append(pre);
    body.append(card);
  }

  const raw = el("details", "raw");
  raw.append(el("summary", null, "Everything recorded for this run (JSON)"));
  raw.append(el("pre", null, JSON.stringify(r, null, 2)));
  body.append(raw);

  block.append(body);
  return block;
}

load();

File: intake-soa/soa/runs.py
"""A local record of every run, from both halves of the assignment.

One SQLite file, one table, stdlib only. The two halves produce different work but answer the same
three questions — how much came out, how much was asked for, and how much still needs a person —
so they share a row shape rather than a table each:

    part='soa'    one protocol PDF extracted   produced=schedules found   flagged=review flags
    part='agent'  one study built on a platform produced=items completed  flagged=escalations

`detail` carries whatever else that run wants to keep, as JSON. Nothing reads it in SQL, so it
never needs a migration when a run learns to record something new.

Write:  soa.server calls log() when a document finishes; `python -m soa.runs ingest <lane.json>…`
        loads the agent's bench output.
Read:   GET /api/runs, rendered at /runs.html.
"""
from __future__ import annotations

import json
import sqlite3
import sys
import uuid
from datetime import datetime, timezone
from pathlib import Path

DB = Path(__file__).resolve().parent.parent / "outputs" / "runs.db"

SCHEMA = """
CREATE TABLE IF NOT EXISTS runs (
  id       TEXT PRIMARY KEY,
  started  TEXT NOT NULL,
  part     TEXT NOT NULL,
  subject  TEXT NOT NULL,
  input    TEXT,
  status   TEXT NOT NULL,
  seconds  REAL,
  produced INTEGER,
  expected INTEGER,
  flagged  INTEGER,
  detail   TEXT
);
CREATE INDEX IF NOT EXISTS runs_started ON runs (started DESC);
"""


def connect(db: Path | None = None) -> sqlite3.Connection:
    path = Path(db or DB)
    path.parent.mkdir(parents=True, exist_ok=True)
    con = sqlite3.connect(path)
    con.row_factory = sqlite3.Row
    con.executescript(SCHEMA)
    return con


def log(part: str, subject: str, status: str, *, input: str = "", seconds: float | None = None,
        produced: int | None = None, expected: int | None = None, flagged: int | None = None,
        detail: dict | None = None, db: Path | None = None) -> str:
    """Record one run. Returns its id. Never raises past the caller: a logging failure must not
    take down the extraction it was recording."""
    run_id = uuid.uuid4().hex[:12]
    try:
        with connect(db) as con:
            con.execute(
                "INSERT INTO runs (id, started, part, subject, input, status, seconds,"
                " produced, expected, flagged, detail) VALUES (?,?,?,?,?,?,?,?,?,?,?)",
                (run_id, datetime.now(timezone.utc).isoformat(timespec="seconds"), part, subject,
                 input, status, seconds, produced, expected, flagged,
                 json.dumps(detail or {}, default=str)),
            )
    except (sqlite3.Error, OSError) as exc:            # locked file, read-only checkout, no disk
        print(f"run log failed: {exc}", file=sys.stderr)
    return run_id


def recent(limit: int = 200, part: str | None = None, db: Path | None = None) -> list[dict]:
    with connect(db) as con:
        rows = con.execute(
            "SELECT * FROM runs" + (" WHERE part = ?" if part else "") + " ORDER BY started DESC LIMIT ?",
            ((part, limit) if part else (limit,)),
        ).fetchall()
    return [{**dict(r), "detail": json.loads(r["detail"] or "{}")} for r in rows]


def totals(db: Path | None = None) -> list[dict]:
    """One summary row per part, for the dashboard header."""
    with connect(db) as con:
        return [dict(r) for r in con.execute(
            "SELECT part, COUNT(*) runs, SUM(produced) produced, SUM(expected) expected,"
            " SUM(flagged) flagged, ROUND(AVG(seconds),1) avg_seconds,"
            " SUM(status != 'done') not_done FROM runs GROUP BY part ORDER BY part")]


# ---------------------------------------------------------------------------
# Part 1 — the agent's bench output, which is written as one JSON file per lane
# ---------------------------------------------------------------------------

def ingest_lane(path: Path, db: Path | None = None) -> str:
    """Load one `intake-agent/bench/out/<lane>.json`. Counts come from the grader in lane.mjs
    (the built study read back and compared with the input file), not from the agent's own tally —
    a run that miscounts itself should show up here as a discrepancy, not be hidden by it."""
    r = json.loads(Path(path).read_text())
    if "score" not in r or "panel" not in r:
        raise ValueError(f"{path} is not a lane result (no score/panel)")
    score = r["score"] or {"seen": {}, "want": {}, "findings": []}
    cards = (r.get("panel") or {}).get("cards") or []
    ledger = (r.get("panel") or {}).get("ledger") or ""
    states: dict[str, int] = {}
    for line in ledger.splitlines():
        parts = line.split("  ·  ")
        if len(parts) > 1:
            states[parts[1]] = states.get(parts[1], 0) + 1

    return log(
        "agent",
        # The file name, not the lane name inside it: two repetitions of one lane
        # are two runs, and telling them apart is the point of a log.
        subject=Path(path).stem,
        input=r.get("ir", ""),
        status="failed" if r.get("error") else "timeout" if r.get("timedOut") else "done",
        seconds=round((r.get("ms") or 0) / 1000, 1),
        produced=score["seen"].get("fields", 0),
        expected=score["want"].get("fields", 0),
        flagged=states.get("escalated", 0),
        detail={
            "url": r.get("url"),
            "panel_status": (r.get("panel") or {}).get("status", ""),
            "visits": f"{score['seen'].get('visits', 0)}/{score['want'].get('visits', 0)}",
            "forms": f"{score['seen'].get('forms', 0)}/{score['want'].get('forms', 0)}",
            "ledger_states": states,
            "findings": score["findings"][:200],
            "gate_cards": [{"signature": c.get("signature", ""), "question": c.get("question", ""),
                            "blocks": c.get("blocks", ""), "tried": c.get("tried", "")} for c in cards],
            "console_errors": r.get("consoleErrors", [])[:20],
        },
        db=db,
    )


def _selfcheck() -> None:
    """Runnable check: a round trip through a scratch database."""
    import tempfile
    with tempfile.TemporaryDirectory() as tmp:
        db = Path(tmp) / "t.db"
        log("soa", "protocol1.pdf", "done", seconds=12.5, produced=1, flagged=2, detail={"a": 1}, db=db)
        log("agent", "mockA-smoke", "done", input="test.ir.json", produced=34, expected=34, flagged=0, db=db)
        rows = recent(db=db)
        assert len(rows) == 2, rows
        assert {r["part"] for r in rows} == {"soa", "agent"}
        assert rows[0]["detail"] == {"a": 1} or rows[1]["detail"] == {"a": 1}
        assert [t["part"] for t in totals(db=db)] == ["agent", "soa"]
        # A bad path must not raise — logging is never allowed to fail a run.
        # The warning it prints next is the check working, not a failure.
        print("  expecting one 'run log failed' warning:")
        assert log("soa", "x", "done", db=Path("/proc/nope/nope.db")) != ""
    print("runs: all checks passed")


if __name__ == "__main__":
    args = sys.argv[1:]
    if args[:1] == ["ingest"]:
        for p in args[1:]:
            print(f"{ingest_lane(Path(p))}  {p}")
    elif args[:1] == ["list"]:
        for r in recent():
            print(f"{r['started']}  {r['part']:6}  {r['subject']:22}  {r['status']:8}"
                  f"  {r['produced']}/{r['expected']}  flagged {r['flagged']}")
    else:
        _selfcheck()
